Downloading media library archive from the export page.

southp opened the pull request:
#### Changes proposed in this Pull Request

This PR adds a "Export Media Library" card to the export page:
![image](https://user-images.githubusercontent.com/1842898/46433284-758e3380-c71e-11e8-8786-51304b8ef782.png)


A user can click on the "Download" button in the card to download an archive of the media library.

In summary, it includes:
1. Expand the `siteSettings.exporter` state tree with `mediaExportData`, including actions, selectors and reducers.
1. Data-layer handlers for interacting with `/rest/v1.1/sites/{id}/exports/media`
1. Add a `<ExportMediaCard>` component.

#### Dependencies
This PR is built on https://github.com/Automattic/wp-calypso/pull/27578.

#### Testing instructions
1. Go to http://calypso.localhost:3000/settings/export/
1. The download button should be disabled at the beginning, waiting for the endpoint response.
1. Once receiving the response, the download button should be enabled and you can click on it to download an archive of your media library.

michelleweber commented:
I might add "entire" or "all the files in," but that's up to you.

What does "as an archive" mean? Will users know what that means? I wonder if we want to be more specific and mention what kind of file they'll actually end up with.

southp commented:
Thanks for taking a look, @michelleweber :)

> What does "as an archive" mean? Will users know what that means? I wonder if we want to be more specific and mention what kind of file they'll actually end up with.

Good question. To be specific, "an archive" here means an [archive file](https://en.wikipedia.org/wiki/Archive_file) in the `tar` format, which can then be extracted on users' computers. Unfortunately I don't really know how to express this in a less tech-savvy way. Could you help me rephrase it? I'd be super grateful.

southp commented:
@michelleweber Thanks for the feedback! I've reworded it and it now looks like this:
![image](https://user-images.githubusercontent.com/1842898/46592131-77773e80-caf2-11e8-8f3d-6d2f51c94610.png)

stephanethomas commented on client/my-sites/exporter/export-media-card/index.js:
```diff
@@ -0,0 +1,57 @@
+	dispatch => ( {
+		recordMediaExportClick: () =>
+			dispatch( recordTracksEvent( 'calypso_export_media_download_button_click' ) ),
+	} )
```
Not mandatory but we should be able to simplify this `dispatch` prop with:

```javascript
{
  recordMediaExportClick: () => recordTracksEvent( 'calypso_export_media_download_button_click' ),
}
```

southp replied:
I always love tips for code slimness. Added in 3c91547 :)

stephanethomas commented on client/state/data-layer/wpcom/sites/exports/media/test/index.js:
```diff
@@ -0,0 +1,65 @@
+// we are mocking impure-lodash here, so that conciergeInitialFetchError() will contain the expected id in the tests
+jest.mock( 'lib/impure-lodash', () => ( {
+	uniqueId: () => 'mock-unique-id',
+} ) );
```
Do we really need that mock? I'm not sure to understand how this test relates to `conciergeInitialFetchError()`.

southp replied:
Ah! Good catch. It's not necessary and the comment was a copy-and-paste error. Fixed in f086733.

southp commented on client/my-sites/exporter/export-media-card/index.js:
```diff
@@ -0,0 +1,57 @@
+		mediaExportUrl: PropTypes.string,
+		recordMediaExportClick: PropTypes.func,
+		translate: PropTypes.func,
+	};
```
Updated in 794066d.

southp commented on client/components/data/query-media-export/index.js:
```diff
@@ -0,0 +1,34 @@
+	render() {
+		return null;
+	}
+}
```
Added in 0db2943.

southp commented on client/my-sites/exporter/export-media-card/index.js:
```diff
@@ -0,0 +1,57 @@
+					headerText={ translate( 'Export media library' ) }
+					mainText={ translate( 'Click the button to download your entire media library.' ) }
+					buttonText={ translate( 'Download' ) }
+					buttonPrimary={ true }
```
TIL! Removed in 61495f4.

southp commented:
@stephanethomas Thanks for the insightful review.

> Note I couldn't apply the server patch as some hunks could not be applied cleanly by the unix 'patch' utility.

Sorry that I didn't update the test plan in advance. It is not necessary anymore since all the backend changes has been deployed. That also means the feature flag, `export-media`, is not necessary anymore, so I've removed it.

When you have some time, please have a revisit to see if I've addressed all your concerns the right way. Thanks :)

southp commented on client/my-sites/exporter/export-card/index.jsx:
```diff
@@ -62,7 +63,12 @@ class ExportCard extends Component {
+											'Or select specific content items to export. Please note this does not include your media library, ' +
+												'but only text content like posts and pages.'
+									  )
+									: translate( 'Or select specific content items to export' ) }
```
Along with removing the feature flag `export-media`, the description only has one variation now.

About:
> A better option could be to add a link to the text that will have the same effect than the chevron icon

If I'm not mistaken, the common practices for this when using `<FoldableCard>` is to set `clickableHeader` as `true` so that the whole card can be clicked to expand. I've also update the sentence as `Or click here to select specific content items to export` to emphasize that it can be clicked. 

Does that work for you? I'm happy to refine the approach, of course :)

southp commented on client/components/data/query-media-export/index.js:
```diff
@@ -0,0 +1,39 @@
+class QueryMediaExport extends Component {
+	static propTypes = {
+		mediaExportUrl: PropTypes.string.isRequired,
+	};
```
Nice catch. Thanks for the eagle eyes 🦅
Updated in d7805cd.

southp commented:
A summary of our discussion:
Since there is currently no easy way to wire expanding of `<FoldableCard>` to the wrapped child components, we will go in two steps:

1. Remove the clickable header prop, implement the proposed new description and ship this first.
1. Address the issue that the options of exporting specific types of content is too obscure in an individual PR.
